lisa/tests/staging/sched_android.py

```diff
@@ -0,0 +1,239 @@
+        )
+
+    @classmethod
+    @abc.abstractmethod
+    def _create_test_bundles(cls, target, res_dir, ftrace_coll):
+        """
+        Collects and returns a ResultBundle per test item.
+        """
+        pass
+
+    @classmethod
+    def _create_test_bundle(cls, target, res_dir, ftrace_coll, item_cls, boost, pi):
+        """
+        Creates and returns a TestBundle for a given item class, and a given
+        schedtune configuration
+        """
+        item_dir = os.path.join(res_dir, 'boost_{}'.format(boost))
+        os.makedirs(item_dir)
+
+        logger = cls.get_logger()
+        logger.info('Running {} with boost={}, prefer_idle={}'.format(
+                    item_cls.__name__, boost, pi))
+        return item_cls.from_target(target, boost, pi, res_dir=item_dir, ftrace_coll=ftrace_coll)
+
+    def _merge_res_bundles(self, res_bundles):
+        """
+        Merge a set of result bundles
+        """
+        overall_bundle = ResultBundle.from_bool(all(res_bundles.values()))
+        for name, bundle in res_bundles.items():
+            overall_bundle.add_metric(name, bundle.metrics)
+
+        overall_bundle.add_metric('failed', [
+            name for name, bundle in res_bundles.items()
+            if bundle.result is Result.FAILED
+        ])
+        return overall_bundle
+
+class SchedTuneFreqItem(SchedTuneItemBase):
+    """
+    Runs a tiny RT rtapp task pinned to a big CPU at a given boost level and
+    checks the frequency selection was performed accordingly.
+    """
+
+    ftrace_conf = FtraceConf({
+        "events" : ["sched_switch", "cpu_frequency"]
+    }, __qualname__)
+
+    @classmethod
+    def get_rtapp_profile(cls, plat_info):
+        cpu = plat_info['capacity-classes'][-1][0]
+        rtapp_profile = {}
+        rtapp_profile['rta_stune'] = Periodic(
+            duty_cycle_pct = 1, # very small task, no impact on freq w/o boost
+            duration_s = 10,
+            period_ms = 16,
+            cpus = [cpu], # pin to big CPU, to focus on frequency selection
+            sched_policy = 'FIFO' # RT tasks have the boost holding feature so
+                                  # the frequency should be more stable, and we
+                                  # shouldn't go to max freq in Android
+        )
+        return rtapp_profile
+
+    def trace_window(self, trace):
+        """
+        Set the boundaries of the trace window to ``cpu_frequency`` events
+        before/after the task's start/end time
+        """
+        rta_start, rta_stop = super().trace_window(trace)
+
+        cpu = self.plat_info['capacity-classes'][-1][0]
+        fdf = trace.df_events('cpu_frequency')
+        fdf = fdf[fdf.cpu == cpu]
+
+        # Find the frequency events before and after the task runs
+        freq_start = fdf[fdf.index < rta_start].index[-1]
+        freq_stop = fdf[fdf.index > rta_stop].index[0]
+
+        return (freq_start, freq_stop)
```

Comment:
What about turning this into something generic and offered by `lisa.trace::Trace`?
Something like:
```python
def get_window(self, event):
```
where you return the window for the first and last specified `event` in `trace`?

Reply:
Just noticed now that, since in `RTATestBundle` we defined `trace` as a `@memoized` `@property` we also end up applying only one time `trace_window()` slicer.

@valschneider perhaps we should call this out better in the documentation of `RTATestBundle::trace`, it's not obvious if you don't read the code or the documentation of `RTATestBundle::trace_window`...

It also means that, if we override the `trace_window` we can do it only one time for each `RTATestBundle` derived class.
@valschneider: are we really happy about the need to create a complete new `RTATestBundle` derived class in case we wanna test different windows for the same trace?

Reply:
> where you return the window for the first and last specified event in trace?

I don't really want that though. I want the `cpu_frequency` event just before the task starts, and the one just after it ends. The rest of the trace doesn't really interest me TBH

Reply:
I don't really see how multiple `TraceView` would work with our flow. In the end we take a single `Trace` (or `TraceView`) and run our tests on it. Why would you want several `TraceView`s?

You can refine a specific view if you want (e.g. use the `super()` one and nest a `TraceView` on it), but that's about it.

This also goes with your @memoized comment - I doesn't make sense to compute a different `trace_window()` since it would use the same slicing code on the same trace (one `RTATestBundle` == one `Trace`).

Reply:
What about cases where you wanna test different conditions on the same trace, which requires different slicing. Don't have examples on hand but I think there could be cases such that.

Reply:
Conceptually that doesn't really fit the current framework. What you could have perhaps is a `RTATestBundle` subclass that take a slicing window as parameter, and have a base class that generates as many as you need (kind of like the `Invariance` thingy in load_tracking).

Reply:
> What about cases where you wanna test different conditions on the same trace, which requires different slicing

The slicing you get through the `trace` property just crops what is outside of rtapp execution. AFAIK, we are never interested by whatever is around. It is possible to have sub-slices of that AFAIK. If we really need to look at what happens before and after rtapp, then it is always possible to override `trace` property with one that would be specific to a test.